[lc709203f] Add battery monitor

ilikecake opened the pull request:
# What does this implement/fix?

Implements the [LC709203F](https://www.adafruit.com/product/4712) battery monitor sensor component. I will also submit a PR to update the documentation with details.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- resolves https://github.com/esphome/feature-requests/issues/2040

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4800

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: lc709203f
    size: 2000
    voltage: 3.7
    battery_voltage:
      name: "Battery Voltage"
    battery_level:
      name: "Battery"
    temperature:
      name: "Pack Temperature"
      thermistor_b_constant: 0xA5A5
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ilikecake committed:
Add an entry for thermistor B-constant

ilikecake committed:
Initial commit

Tested and seems to have basic functionality. Needs cleanup and error checking.

ilikecake committed:
Add comments and implement voltage setting.

ilikecake committed:
Initial commit

ilikecake committed:
Change polling time to 60sec.

ilikecake committed:
Remove leftover debug line.

ilikecake committed:
Remove comment lines

ilikecake committed:
Updates for flake8

ilikecake committed:
Updates for black checks

ilikecake committed:
Remove trailing whitespace.

ilikecake committed:
Remove more whitespace

ilikecake committed:
Add spaces between comment symbol and comment.

ilikecake committed:
remove spaces on constants.

ilikecake committed:
Change indent to two spaces from four.

ilikecake committed:
Fix indenting.

ilikecake committed:
Clang formatting fixes

ilikecake committed:
More clang formatting fixes.

ilikecake committed:
Even more clang formatting updates.

ilikecake committed:
Merge branch 'dev' into lc709203f

ilikecake committed:
Run build_codeowners script

ilikecake committed:
Change variable formatting

ilikecake committed:
Fix variable formatting

ilikecake committed:
Fix more variable formatting.

ilikecake committed:
Remove extra return statements.

ilikecake committed:
Fix capitalization.

ilikecake committed:
Clang formatting fixes.

ilikecake committed:
Remove unneeded initialization.

ilikecake committed:
Correct TAG definition.

ilikecake committed:
Merge branch 'dev' into lc709203f

ilikecake committed:
Merge branch 'dev' into lc709203f

ilikecake committed:
Merge branch 'dev' into lc709203f

swoboda1337 commented on tests/components/lc709203f/test.esp32-ard.yaml:
```diff
@@ -0,0 +1,16 @@
+    scl: 16
+    sda: 17
+
+sensor:
```
Tests need to be refactored. Should have one common file and then substitutions for pins. Look at the current tests for examples.

ilikecake replied:
Thanks for the review. Can you point me in the direction of an example of the current tests? I tried to follow the formatting from some of the other tests.

swoboda1337 replied:
> Thanks for the review. Can you point me in the direction of an example of the current tests? I tried to follow the formatting from some of the other tests.

https://github.com/esphome/esphome/tree/dev/tests/components/am2315c

You will see there is now a common and all tests include that common. This was changed not too long ago.

swoboda1337 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,328 @@
+          return;
+      }
+  }
+  */
```
Why have dead/commented code, its not formatted correctly either.

swoboda1337 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,328 @@
+
+void Lc709203f::set_pack_size(uint16_t pack_size) {
+  uint16_t pack_size_array[6] = {100, 200, 500, 1000, 2000, 3000};
+  uint16_t apa_array[6] = {0x08, 0x0B, 0x10, 0x19, 0x2D, 0x36};
```
These two should be static const?

swoboda1337 commented on esphome/components/lc709203f/sensor.py:
```diff
@@ -0,0 +1,98 @@
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(lc709203f),
+            cv.Optional(CONF_SIZE, default="500"): cv.int_range(200, 3000),
```
So you cant set values bewteen100mAh and 200mAh??

ilikecake replied:
Good catch. That should allow values between 100 and 3000. TBH, even pack sizes above that range are probably okay, but the datasheet only provides calibration data for that range of sizes.

swoboda1337 commented on esphome/components/lc709203f/sensor.py:
```diff
@@ -0,0 +1,98 @@
+
+    if CONF_BATTERY_LEVEL in config:
+        sens = await sensor.new_sensor(config[CONF_BATTERY_LEVEL])
+        cg.add(var.set_battery_remaining_sensor(sens))
```
Why is this if not consistent with the other two?
```
if CONF_BATTERY_LEVEL in config:
``` 
Vs 
```
if voltage_config := config.get(CONF_BATTERY_VOLTAGE):
```

ilikecake replied:
I suspect that is a remnant of the examples that I was using to write this. Is one of those methods preferred?

swoboda1337 replied:
Just make it consistent

kbx81 replied:
[The walrus is preferred.](https://developers.esphome.io/contributing/code/#python) Please change them all to use the walrus operator.

swoboda1337 commented on esphome/const.py:
```diff
@@ -874,6 +874,7 @@
 CONF_TEXT = "text"
 CONF_TEXT_SENSORS = "text_sensors"
 CONF_THEN = "then"
+CONF_THERMISTOR_B_CONSTANT = "thermistor_b_constant"
```
Not sure its worth adding this to const can just define this in your python file. It is unlikely to be reused by anyone else.

ilikecake committed:
Correct formatting of tests.

ilikecake committed:
Remove commented code that probably wouldn't work.

ilikecake committed:
Make the APA lookup table 'static const'

ilikecake committed:
Fixes to the python files based on review

-Moved CONF_THERMISTOR_B_CONSTANT into my sensor.py file instead of the const.py
-Corrected pack size limits
-Standardize the sensor setup functions.

ilikecake committed:
Add newlines at the end of the files.

ilikecake committed:
Make static const variables upper case

ilikecake committed:
Merge branch 'dev' into lc709203f

ilikecake commented:
@swoboda1337, thanks for looking at this. I have made updates based on your review. Please let me know if you have any other comments.

ilikecake commented:
Note that this PR would close this [issue](https://github.com/esphome/feature-requests/issues/2040).

swoboda1337 commented:
> N/A

If you add `- resolves https://github.com/esphome/feature-requests/issues/2040` to `Related issue or feature (if applicable):` it will link them and auto close the issue when merged.

ilikecake committed:
Merge branch 'dev' into lc709203f

kbx81 commented on esphome/components/lc709203f/sensor.py:
```diff
@@ -0,0 +1,99 @@
+                entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+            ).extend(
+                {
+                    cv.Optional(
+                        CONF_THERMISTOR_B_CONSTANT, default="0x0D34"
+                    ): cv.int_range(0, 0xFFFF),
```
Any reason this is in hex? Seems a bit...weird...here. 😄

ilikecake replied:
Hmm, not really. I just matched how they called out the default in the datasheet. Should I make this a required variable? If the user is using a thermistor, this value needs to be set, and I don't really know if that constant would be suitable for any generic thermistor. It might be better to make that value required if the user is using the thermistor.

kbx81 replied:
For the [NTC sensor](https://github.com/esphome/esphome/blob/1c5dc63eb4df01935d19652ab5e9b74dbe329486/esphome/components/ntc/sensor.py#L92), it is required.
I don't think we should assume any defaults here if the user is providing this part.

kbx81 replied:
...which reminds me -- please use the constant from `const.py` here.

ilikecake replied:
I will change it to required. I don't see a config variable in const.py for this, which one do you mean? Should I use the same `CONF_B_CONSTANT = "b_constant"` from the NTC sensor you linked?

kbx81 replied:
> Should I use the same `CONF_B_CONSTANT = "b_constant"` from the NTC sensor you linked?

Yes please. 🙂

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  LOG_UPDATE_INTERVAL(this);
+  ESP_LOGCONFIG(TAG, "  Pack Size: %d mAH", this->pack_size_);
+  ESP_LOGCONFIG(TAG, "  Pack APA: 0x%02X", this->apa_);
+  if (this->pack_voltage_ == 0x0000) {
+    ESP_LOGCONFIG(TAG, "  Pack Rated Voltage: 3.8V");
+  } else {
+    // This is only true if the pack_voltage_ is 0x0001. The config validator should
+    //  have already made sure that this value can be only 0 or 1.
+    ESP_LOGCONFIG(TAG, "  Pack Rated Voltage: 3.7V");
+  }
```
Instead of storing two copies of the same string:
```suggestion
  // This is only true if the pack_voltage_ is 0x0001. The config validator should
  //  have already made sure that this value can be only 0 or 1.
  ESP_LOGCONFIG(TAG, "  Pack Rated Voltage: 3.%sV", this->pack_voltage_ == 0x0000 ? "8" : "7");
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  read_buffer[1] = register_to_read;
+  read_buffer[2] = ((this->address_) << 1) | 0x01;
+
+  for (int i = 0; i <= LC709203F_I2C_RETRY_COUNT; i++) {
```
```suggestion
  for (uint8_t i = 0; i <= LC709203F_I2C_RETRY_COUNT; i++) {
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  write_buffer[3] = (value_to_set >> 8) & 0xFF;  // High byte
+  write_buffer[4] = this->crc8_(write_buffer, 4);
+
+  for (int i = 0; i <= LC709203F_I2C_RETRY_COUNT; i++) {
```
```suggestion
  for (uint8_t i = 0; i <= LC709203F_I2C_RETRY_COUNT; i++) {
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  //  does not record power usage. If there is significant power consumption during sleep mode,
+  //  the pack RSOC will likely no longer be correct. Because of that, I do not implement
+  //  sleep mode on this device.
+  if (this->set_register_(LC709203F_IC_POWER_MODE, LC709203F_POWER_MODE_ON) == i2c::NO_ERROR) {
+    if (this->set_register_(LC709203F_APA, this->apa_) == i2c::NO_ERROR) {
+      if (this->set_register_(LC709203F_CHANGE_OF_THE_PARAMETER, this->pack_voltage_) == i2c::NO_ERROR) {
+        this->state_ = LC709203F_STATE_RSOC;
+      }
+    }
+  }
```
Nothing specifically wrong here, but...how many levels deep are you going to stack these? 😅
```suggestion
  if (this->set_register_(LC709203F_IC_POWER_MODE, LC709203F_POWER_MODE_ON) != i2c::NO_ERROR) {
    return;
  }
  if (this->set_register_(LC709203F_APA, this->apa_) != i2c::NO_ERROR) {
    return;
  }
  if (this->set_register_(LC709203F_CHANGE_OF_THE_PARAMETER, this->pack_voltage_) != i2c::NO_ERROR) {
    return;
  }
  this->state_ = LC709203F_STATE_RSOC;
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  //  call to update().
+  ESP_LOGCONFIG(TAG, "Setting up LC709203F...");
+
+  this->state_ = LC709203F_STATE_INIT;
```
Why not just initialize this in the header?

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  } else if (this->state_ == LC709203F_STATE_INIT) {
+    // We should only get here if the init sequence failed during the setup() function.
+    //  This would likely occur because of a repeated failure on the I2C bus.
+    if (this->set_register_(LC709203F_IC_POWER_MODE, LC709203F_POWER_MODE_ON) == i2c::NO_ERROR) {
+      if (this->set_register_(LC709203F_APA, this->apa_) == i2c::NO_ERROR) {
+        if (this->set_register_(LC709203F_CHANGE_OF_THE_PARAMETER, this->pack_voltage_) == i2c::NO_ERROR) {
+          this->state_ = LC709203F_STATE_RSOC;
+        }
+      }
+    }
```
Same thought as I have in `setup()`

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+          this->state_ = LC709203F_STATE_NORMAL;
+        }
+      }
+    } else {
+      // The device expects to get updates to the temperature in this mode.
+      //  I am not doing that now. The temperature register defaults to 25C.
+      //  In theory, we could have another temperature sensor and have ESPHome
+      //  send updated temperature to the device occasionally, but I have no idea
+      //  how to make that happen.
+      if (this->set_register_(LC709203F_STATUS_BIT, 0x0000) == i2c::NO_ERROR) {
+        this->state_ = LC709203F_STATE_NORMAL;
+      }
+    }
```
```suggestion
    } else if (this->set_register_(LC709203F_STATUS_BIT, 0x0000) == i2c::NO_ERROR) {
      // The device expects to get updates to the temperature in this mode.
      //  I am not doing that now. The temperature register defaults to 25C.
      //  In theory, we could have another temperature sensor and have ESPHome
      //  send updated temperature to the device occasionally, but I have no idea
      //  how to make that happen.
      this->state_ = LC709203F_STATE_NORMAL;
    }
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+      snprintf(this->error_code_buffer_, 50, "Error code %d when reading from register 0x%02X", return_code,
+               register_to_read);
+      this->status_set_warning(this->error_code_buffer_);
+    } else {
+      if (this->crc8_(read_buffer, 5) != read_buffer[5]) {
```
Another if/else I think you can combine. 🙂

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  uint8_t crc = 0x00;
+  const uint8_t polynomial(0x07);
+
+  for (int j = length_of_crc; j; --j) {
+    crc ^= *byte_buffer++;
+
+    for (int i = 8; i; --i) {
```
```suggestion
  for (uint8_t j = length_of_crc; j; --j) {
    crc ^= *byte_buffer++;

    for (uint8_t i = 8; i; --i) {
```

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+
+  // The size is used to calculate the 'Adjustment Pack Application' number.
+  // Here we assume a type 01 or type 03 battery and do a linear curve fit to find the APA.
+  for (int i = 0; i < sizeof(PACK_SIZE_ARRAY) / sizeof(PACK_SIZE_ARRAY[0]); i++) {
```
I think you can use `uint8_t` here, too, since, if I'm understanding this correctly, you're basically just doing `for (uint8_t i = 0; i < 6; i++)`...
```suggestion
  for (uint8_t i = 0; i < sizeof(PACK_SIZE_ARRAY) / sizeof(PACK_SIZE_ARRAY[0]); i++) {
```

ilikecake committed:
Simplify to remove redundant strings.

ilikecake committed:
Change int to uint8_t in various for loops.

ilikecake committed:
Combine if/else statements.

ilikecake committed:
Remove nested if statements.

ilikecake committed:
Move setting initial state to header.

ilikecake committed:
Combine else/if statements.

Co-authored-by: Keith Burzinski <kbx81x@gmail.com>

ilikecake committed:
Remove trailing whitespace.

ilikecake committed:
Fix spacing

ilikecake committed:
Hardcode the length of the APA array.

ilikecake committed:
Merge branch 'dev' into lc709203f

ilikecake committed:
Change the config so that the B constant is required if you are using a temp sensor.

kbx81 commented on esphome/components/lc709203f/lc709203f.h:
```diff
@@ -0,0 +1,57 @@
+static const uint8_t LC709203F_STATE_INIT = 0x01;
+static const uint8_t LC709203F_STATE_RSOC = 0x02;
+static const uint8_t LC709203F_STATE_TEMP_SETUP = 0x03;
+static const uint8_t LC709203F_STATE_NORMAL = 0x00;
```
Any reason these aren't an `enum`? That might make more sense since you're tracking state -- there should only be one state at a time.

kbx81 commented on esphome/components/lc709203f/lc709203f.h:
```diff
@@ -0,0 +1,57 @@
+  uint8_t state_ = LC709203F_STATE_INIT;
+  uint16_t pack_voltage_;
+
+  // A buffer to store error code messages. We put this here so as to not have to
+  //  duplicate buffers in the different functions
+  char error_code_buffer_[50];
```
Don't do this here. You're tying up 50 bytes of RAM per instance of the component all the time. You really only need this memory when you're printing an error which I see being done in only two methods. Just allocate it there as needed.

ilikecake replied:
Sorry, I am used to my microcontrollers having a static memory map. If I move the declaration of that array to the start of the `get_register_` function, does that help? Wont it still get allocated every time the function is called? What is the preferred way to handle this?

kbx81 replied:
Just do something like [this](https://github.com/esphome/esphome/blob/9dd404598464d103392c29d13e946e65fd2dedd3/esphome/components/ch422g/ch422g.cpp#L97)

kbx81 commented on esphome/components/lc709203f/lc709203f.h:
```diff
@@ -0,0 +1,57 @@
+  void set_pack_size(uint16_t pack_size);
+  void set_thermistor_b_constant(uint16_t b_constant);
+  void set_pack_voltage(LC709203FBatteryVoltage pack_voltage);
+  void set_voltage_sensor(sensor::Sensor *voltage_sensor) { voltage_sensor_ = voltage_sensor; }
+  void set_battery_remaining_sensor(sensor::Sensor *battery_remaining_sensor) {
+    battery_remaining_sensor_ = battery_remaining_sensor;
+  }
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
```
Please ensure you have the `this->` prefix on all member variable accesses & calls to class methods.
```suggestion
  void set_voltage_sensor(sensor::Sensor *voltage_sensor) { this->voltage_sensor_ = voltage_sensor; }
  void set_battery_remaining_sensor(sensor::Sensor *battery_remaining_sensor) {
    this->battery_remaining_sensor_ = battery_remaining_sensor;
  }
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
```

kbx81 commented:
Bumping https://github.com/esphome/esphome/pull/8037/files#r2092316064

ilikecake committed:
Update name of B-constant config variable

Update config variable to match what is used in the NTC sensor.

ilikecake committed:
Use walrus operator in sensor setup function.

ilikecake committed:
Add missing -> operator to setters

Co-authored-by: Keith Burzinski <kbx81x@gmail.com>

ilikecake committed:
Use an enum for the device states.

ilikecake committed:
Merge branch 'lc709203f' of https://github.com/ilikecake/esphome into lc709203f

ilikecake committed:
Formatting fixes

ilikecake committed:
Remove fixed buffer for error messages.

ilikecake committed:
Rename state variable

ilikecake committed:
Formatting fixes.

kbx81 committed:
Appease CI

kbx81 committed:
Formatting

kbx81 committed:
Fix some floats

kbx81 committed:
Merge branch 'dev' into lc709203f

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+  //  initialization code checks the return code from those functions. If they don't return
+  //  NO_ERROR (0x00), that part of the initialization aborts and will be retried on the next
+  //  call to update().
+  ESP_LOGCONFIG(TAG, "Setting up LC709203F...");
```
```suggestion
  ESP_LOGCONFIG(TAG, "Running setup");
```

kbx81 committed:
Update esphome/components/lc709203f/lc709203f.cpp

kbx81 commented on esphome/components/lc709203f/lc709203f.cpp:
```diff
@@ -0,0 +1,302 @@
+    // Nothing in this code actually sets the device to failed, so this should never trigger.
+    //  I am leaving this in incase I want to implement a check during init to verify the
+    //  correct device is on the bus.
+    ESP_LOGCONFIG(TAG, "  I2C Device at 0x0B does not appear to be a LC709203F");
```
```suggestion
    ESP_LOGCONFIG(TAG, ESP_LOG_MSG_COMM_FAIL);
```

kbx81 replied:
@ilikecake will you implement this check now? If not, let's remove this block of code and I'll merge the PR. You can always do a follow-up to add enhancements.

ilikecake replied:
I don't know a good way to check this for this device. If there were some sort of ID register that we could read to verify the device is what we think, that would be great. In this case, I don't think there is any reliable way to do that. There is an 'IC Version' register, but it is unclear if that is specific to an individual chip or the same for every chip. I will remove this code for now.

ilikecake replied:
On second thought, while I don't think there is a good way to verify device ID during init, there might be a use for `mark_failed()`. The device can get confused if you poll it too often and respond incorrectly over i2c. That is why I implement 10 retries and then return an error code during register read and write operations. The way the code works now, the higher level functions will ignore i2c transactions that don't return the `NO_ERROR` code, but they will try again at the next poll time. Do you think it would be worthwhile to use `mark_failed()` if the device has 10 failed i2c transactions in a row? FWIW, I have never seen more than 2-3 failed transactions consecutively in my testing.

kbx81 replied:
I think, in general, we only use `mark_failed()` for unrecoverable failures. If you can recover communication by retrying reads/writes, then there's probably no reason to `mark_failed()` and instead maybe just set the "warning" status/flag.

ilikecake replied:
ok, that makes sense. The code will already send warnings on i2c bus errors, so I removed that block of code.

kbx81 committed:
Update esphome/components/lc709203f/lc709203f.cpp

ilikecake committed:
Remove check for is_failed()

kbx81 reviewed (approved):
LGTM -- thanks for working on this! 🍻
